_posts/2019-04-24-lecture-27.md

```diff
@@ -0,0 +1,434 @@
+
+- **Bulk synchronous parallel:** In this scheme, all the worker nodes get a fresh copy of the parameters from the master, calculate the updates using these parameters and send the updates to the master. The master then waits till it has received an update from all the nodes and then updates the parameters. It then broadcasts the new parameters to all the nodes. This is slowed down by the fact that the nodes have to wait for all other nodes to finish to receive a new copy.
+- **Asynchronous parallel:** Essentially Hogwild!. The worker nodes do their jobs without waiting for the slower nodes to catch up. 
+- **Stale synchronous/bounded asynchronous parallel:** A middle ground between bulk synchronous and asynchronous schemes. Unlike the bulk synchronous scheme, the worker nodes do not wait for all the other nodes to have calculated an update, but unlike the asynchronous scheme, they're not allowed to get too far ahead of the slower nodes either. In particular, there is a limit (staleness threshold) to the number of steps a particular worker can be ahead by, after which it must wait for the slower threads to catch up.
+
+Empirically, it has been shown that stale synchronous parallel updates converge faster (in seconds) than the other two.
+
+### Sequential Coordinate Descent
+
+Another popular optimization technique that is commonly used is the Coordinate Descent Algorithm. It works as follows:
+
+- Randomly pick a single parameter $\theta_i$ from the vector of parameters $\theta$
+- Given the loss function $\mathcal{L}(\theta)$, solve the equation $\frac{\partial \mathcal{L}(\theta)}{\partial\theta_i} = 0$ and update the parameter $\theta_i$ to the solution
+- Repeat
+
+This algorithm suits Lasso regression very well as the partial derivative turns out to be a linear function of $\theta_i$, and therefore the equation is easily solvable.
+
+#### Block Coordinate Descent:
+
+Coordinate Descent lends itself to a simple extension called Block Coordinate Descent. Instead of picking a single parameter in the first step, we select a subset of parameters. In the second step, we solve a system of equations by obtained by setting the partial derivatives with respect to all the parameters in the subset to zero.
+
+### Shotgun, a Parallel Coordinate Descent Algorithm <d-cite key="bradley2011parallel"></d-cite>
+
+- An algorithm very similar to Hogwild!, in that it is simple to implement.
+- Each worker randomly picks a coordinate (a single parameter) to update.
+- The updates are done in parallel with no locks.
+- If the features are nearly independent, it scales almost linearly with the number of machines.
+- As can be seen in the image below, oncurrent updates to coordinates often overshoot if the features are correlated and may slow down the covergence, or in the worst case, cause a divergence.
+
+<img src="{{ '/assets/img/notes/lecture-27/pcsd.png' | relative_url }}" />
+
+### Block-greedy Coordinate Descent <d-cite key="scherrer2012feature"></d-cite>
```

Comment:
Please format this section in paragraph form so that it reads more like a blog post, rather than bullet points from lecture slides. Here's an example that I liked: https://sailinglab.github.io/pgm-spring-2019/notes/lecture-13/